src/Library/WebServer.cs

```diff
@@ -0,0 +1,144 @@
+﻿namespace Microsoft.IstioMixerPlugin.Library
+{
+    using Microsoft.IstioMixerPlugin.Common;
+    using Microsoft.IstioMixerPlugin.Library.Inputs;
+    using System;
+    using System.Net;
+    using System.Runtime.Serialization.Json;
+
+    public class WebServer
+    {
+        private HttpListener listener;
+        private volatile bool isRunning = false;
+        private readonly Configuration config;
+        private readonly Object syncObject = new Object();
+        internal bool IsRunning
+        {
+            get
+            {
+                return this.isRunning;
+            }
+        }
+
+        public WebServer(string config)
+        {
+            if (!HttpListener.IsSupported)
+            {
+                string logLine = FormattableString.Invariant($"HttpListener is not supported");
+                Diagnostics.LogError(logLine);
+                throw new InvalidOperationException(logLine);
+            }
+
+            if (config == null || config.Length == 0)
+            {
+                string logLine = FormattableString.Invariant($"config missing");
+                Diagnostics.LogError(logLine);
+                throw new ArgumentException(logLine);
+            }
+
+            this.config = new Configuration(config);
+        }
+
+        public void Start()
+        {
+            if (this.isRunning)
+            {
+                Diagnostics.LogInfo("WebServer already running .. exiting");
+                return;
+            }
+            // Create a listener.
+            this.listener = new HttpListener();
+            this.listener.Prefixes.Add(this.config.HttpPrefix);
+
+            lock (this.syncObject)
+            {
+                try
+                {
+                    this.listener.Start();
+                    listener.BeginGetContext(new AsyncCallback(this.ListenerCallbackAsync), this.listener);
+                    this.isRunning = true;
+                }
+                catch (Exception e)
+                {
+                    Diagnostics.LogError(e.ToString());
+                    // should not blow up if it cannot run the webserver
+                }
+            }
+
+            Diagnostics.LogInfo(FormattableString.Invariant($"Webserver running"));
+        }
+
+        public void Stop()
+        {
+            lock (this.syncObject)
+            {
+                if (!this.isRunning)
```

Comment:
This is correct, Start() should be doing the same thing. Another check on the outside of lock() could make it faster, but we don't really care since we don't expect a saturation here.